In [9]:
%load_ext autoreload
%autoreload 2

In [121]:
import pandas as pd
from decompose_packet import *
from ids_mapping import load_ids

**Load packets from local csv for test purpose**

In [72]:
csv = pd.read_csv('./data/example_packets.csv')
list_packets = csv['data'].tolist()

**Load id mapping from local dic**

In [16]:
dict_ids = load_ids()

In [40]:
class color:
   PURPLE = '\033[95m'
   CYAN = '\033[96m'
   DARKCYAN = '\033[36m'
   BLUE = '\033[94m'
   GREEN = '\033[92m'
   YELLOW = '\033[93m'
   RED = '\033[91m'
   BOLD = '\033[1m'
   UNDERLINE = '\033[4m'
   END = '\033[0m'

In [73]:
for packet in list_packets:
    
    id_, packet_data_bit = main(payload=packet)
#     print(color.BOLD + dict_ids[str(id_)] + color.END)
    
    print("\n")
    print("------ NEXT PACKET ------")

 Start decompiling the packet 01:a4:00:00:03:0e 
***
  Split header and body 
The header is 01:a4
The body is 00:00:03:0e
***
  Convert header in bin 
Converted byte in 01a4
Header in binary is 0000000110100100
***
  Extract the id and the message size's size in binary 
The id in binary is 00000001101001
The message size's size in binary is 00
***
  Convert the id and the message size's size from binary to int 
The id is 105 which corresponds to the  GameMapMovementConfirmMessage  action 
The message size's is onto 0 byte(s)
***
  Break the body into the size of the message and the packet's data 
The message does not contain any data
The packet data is 00:00:03:0e


------ NEXT PACKET ------
 Start decompiling the packet 48:91:02:4d:69:93:9c 
***
  Split header and body 
The header is 48:91
The body is 02:4d:69:93:9c
***
  Convert header in bin 
Converted byte in 4891
Header in binary is 0100100010010001
***
  Extract the id and the message size's size in binary 
The id in binary is 01

# Utils function

In [213]:
def transform_list_to_bytes_array(list_bytes):
    return ':'.join(list_bytes)

## Deserialise a packet (GameContextRefreshEntityLookMessage)

In [214]:
packet = "3c:21:57:42:66:63:73:00:24:a0:00:ff:04:00:00:00:04:01:dc:14:3c:02:ff:ff:f0:03:ff:ff:f0:04:76:64:43:00:01:78:00:01:02:00:02:00:06:50:d1:10:a8:02:e8:17:fb:1b:f3:03:00:07:01:ff:d3:83:02:ff:ff:ff:03:ff:fe:db:04:ff:ff:ff:05:00:00:00:07:00:00:00:08:ff:ff:ff:00:01:8c:01:00:00:2e:d5:0b:42:66:63:73:00:24:a0:00:00:08:00:3c:21:5c:42:4c:01:b1:80:92:80:00:ff:04:00:01:a8:1a:00:04:01:6a:48:00:02:6a:48:00:03:6a:48:00:04:73:60:42:00:01:78:00:02:06:00:a9:01:00:00:00:00:00:00:00:00:02:00:02:00:06:6e:ff:10:2b:e3:02:e2:02:de:05:00:05:01:ff:da:95:02:2d:2d:2d:03:27:4e:5c:04:23:47:51:05:2f:2f:2f:00:01:91:01:00:00:2e:d5:0b:42:4c:01:b1:80:92:80:00:00:08:00:3c:21:3d:42:66:6c:2b:c0:24:a0:00:8a:13:00:00:00:00:00:00:00:01:02:00:02:00:06:1f:85:10:f4:01:92:01:81:0c:9b:08:00:05:01:ea:c0:62:02:f3:e0:a4:03:00:00:00:04:ff:ff:ff:05:ad:8d:39:00:01:5f:00:00:2e:d5:0b:42:66:6c:2b:c0:24:a0:00:00:08:00"

In [215]:
id_, packet_data = main(packet)
bytes_list = packet_data.split(':')

 Start decompiling the packet 3c:21:57:42:66:63:73:00:24:a0:00:ff:04:00:00:00:04:01:dc:14:3c:02:ff:ff:f0:03:ff:ff:f0:04:76:64:43:00:01:78:00:01:02:00:02:00:06:50:d1:10:a8:02:e8:17:fb:1b:f3:03:00:07:01:ff:d3:83:02:ff:ff:ff:03:ff:fe:db:04:ff:ff:ff:05:00:00:00:07:00:00:00:08:ff:ff:ff:00:01:8c:01:00:00:2e:d5:0b:42:66:63:73:00:24:a0:00:00:08:00:3c:21:5c:42:4c:01:b1:80:92:80:00:ff:04:00:01:a8:1a:00:04:01:6a:48:00:02:6a:48:00:03:6a:48:00:04:73:60:42:00:01:78:00:02:06:00:a9:01:00:00:00:00:00:00:00:00:02:00:02:00:06:6e:ff:10:2b:e3:02:e2:02:de:05:00:05:01:ff:da:95:02:2d:2d:2d:03:27:4e:5c:04:23:47:51:05:2f:2f:2f:00:01:91:01:00:00:2e:d5:0b:42:4c:01:b1:80:92:80:00:00:08:00:3c:21:3d:42:66:6c:2b:c0:24:a0:00:8a:13:00:00:00:00:00:00:00:01:02:00:02:00:06:1f:85:10:f4:01:92:01:81:0c:9b:08:00:05:01:ea:c0:62:02:f3:e0:a4:03:00:00:00:04:ff:ff:ff:05:ad:8d:39:00:01:5f:00:00:2e:d5:0b:42:66:6c:2b:c0:24:a0:00:00:08:00 
***
  Split header and body 
The header is 3c:21
The body is 57:42:66:63:73:00:24:a0:00:ff:04:00

In [216]:
transform_list_to_bytes_array(bytes_list)

'42:66:63:73:00:24:a0:00:ff:04:00:00:00:04:01:dc:14:3c:02:ff:ff:f0:03:ff:ff:f0:04:76:64:43:00:01:78:00:01:02:00:02:00:06:50:d1:10:a8:02:e8:17:fb:1b:f3:03:00:07:01:ff:d3:83:02:ff:ff:ff:03:ff:fe:db:04:ff:ff:ff:05:00:00:00:07:00:00:00:08:ff:ff:ff:00:01:8c:01:00:00:2e:d5:0b:42:66:63:73:00:24:a0:00:00:08:00:3c:21:5c:42:4c:01:b1:80:92:80:00:ff:04:00:01:a8:1a:00:04:01:6a:48:00:02:6a:48:00:03:6a:48:00:04:73:60:42:00:01:78:00:02:06:00:a9:01:00:00:00:00:00:00:00:00:02:00:02:00:06:6e:ff:10:2b:e3:02:e2:02:de:05:00:05:01:ff:da:95:02:2d:2d:2d:03:27:4e:5c:04:23:47:51:05:2f:2f:2f:00:01:91:01:00:00:2e:d5:0b:42:4c:01:b1:80:92:80:00:00:08:00:3c:21:3d:42:66:6c:2b:c0:24:a0:00:8a:13:00:00:00:00:00:00:00:01:02:00:02:00:06:1f:85:10:f4:01:92:01:81:0c:9b:08:00:05:01:ea:c0:62:02:f3:e0:a4:03:00:00:00:04:ff:ff:ff:05:ad:8d:39:00:01:5f:00:00:2e:d5:0b:42:66:6c:2b:c0:24:a0:00:00:08:00'

## **general deserialisation function**

      public function deserializeAs_GameContextRefreshEntityLookMessage(input:ICustomDataInput) : void
      {
         this._idFunc(input);
         this.look = new EntityLook();
         this.look.deserialize(input);
      }

1) **_idFunc**

      private function _idFunc(input:ICustomDataInput) : void
      {
         this.id = input.readDouble();
         if(this.id < -9007199254740992 || this.id > 9007199254740992)
         {
            throw new Error("Forbidden value (" + this.id + ") on element of GameContextRefreshEntityLookMessage.id.");
         }
      }

In [217]:
bytes_to_read = bytes_list

In [218]:
# equivalent to readdouble
def read_double(bytes_):
    # reads 64 bits / 8 bytes
    
    res_hex = ":".join(bytes_[:8])
    res = convert_byte_to_int(res_hex)
    
    new_bytes_ = bytes_[8:]
    
    return res, new_bytes_

In [219]:
res, bytes_to_read = read_double(bytes_to_read)

**this.is will take the value of res which is 4784620999702519808 in our example then execute if statement**

In [220]:
(res < -9007199254740992) or (res > 9007199254740992)

True

In [221]:
transform_list_to_bytes_array(bytes_list)

'42:66:63:73:00:24:a0:00:ff:04:00:00:00:04:01:dc:14:3c:02:ff:ff:f0:03:ff:ff:f0:04:76:64:43:00:01:78:00:01:02:00:02:00:06:50:d1:10:a8:02:e8:17:fb:1b:f3:03:00:07:01:ff:d3:83:02:ff:ff:ff:03:ff:fe:db:04:ff:ff:ff:05:00:00:00:07:00:00:00:08:ff:ff:ff:00:01:8c:01:00:00:2e:d5:0b:42:66:63:73:00:24:a0:00:00:08:00:3c:21:5c:42:4c:01:b1:80:92:80:00:ff:04:00:01:a8:1a:00:04:01:6a:48:00:02:6a:48:00:03:6a:48:00:04:73:60:42:00:01:78:00:02:06:00:a9:01:00:00:00:00:00:00:00:00:02:00:02:00:06:6e:ff:10:2b:e3:02:e2:02:de:05:00:05:01:ff:da:95:02:2d:2d:2d:03:27:4e:5c:04:23:47:51:05:2f:2f:2f:00:01:91:01:00:00:2e:d5:0b:42:4c:01:b1:80:92:80:00:00:08:00:3c:21:3d:42:66:6c:2b:c0:24:a0:00:8a:13:00:00:00:00:00:00:00:01:02:00:02:00:06:1f:85:10:f4:01:92:01:81:0c:9b:08:00:05:01:ea:c0:62:02:f3:e0:a4:03:00:00:00:04:ff:ff:ff:05:ad:8d:39:00:01:5f:00:00:2e:d5:0b:42:66:6c:2b:c0:24:a0:00:00:08:00'

In [222]:
transform_list_to_bytes_array(bytes_to_read)

'ff:04:00:00:00:04:01:dc:14:3c:02:ff:ff:f0:03:ff:ff:f0:04:76:64:43:00:01:78:00:01:02:00:02:00:06:50:d1:10:a8:02:e8:17:fb:1b:f3:03:00:07:01:ff:d3:83:02:ff:ff:ff:03:ff:fe:db:04:ff:ff:ff:05:00:00:00:07:00:00:00:08:ff:ff:ff:00:01:8c:01:00:00:2e:d5:0b:42:66:63:73:00:24:a0:00:00:08:00:3c:21:5c:42:4c:01:b1:80:92:80:00:ff:04:00:01:a8:1a:00:04:01:6a:48:00:02:6a:48:00:03:6a:48:00:04:73:60:42:00:01:78:00:02:06:00:a9:01:00:00:00:00:00:00:00:00:02:00:02:00:06:6e:ff:10:2b:e3:02:e2:02:de:05:00:05:01:ff:da:95:02:2d:2d:2d:03:27:4e:5c:04:23:47:51:05:2f:2f:2f:00:01:91:01:00:00:2e:d5:0b:42:4c:01:b1:80:92:80:00:00:08:00:3c:21:3d:42:66:6c:2b:c0:24:a0:00:8a:13:00:00:00:00:00:00:00:01:02:00:02:00:06:1f:85:10:f4:01:92:01:81:0c:9b:08:00:05:01:ea:c0:62:02:f3:e0:a4:03:00:00:00:04:ff:ff:ff:05:ad:8d:39:00:01:5f:00:00:2e:d5:0b:42:66:6c:2b:c0:24:a0:00:00:08:00'

2. **look.deserialize**

      public function deserializeAs_EntityLook(input:ICustomDataInput) : void
      {
         var _val2:uint = 0;
         var _val3:int = 0;
         var _val4:int = 0;
         var _item5:SubEntity = null;
         this._bonesIdFunc(input);
         var _skinsLen:uint = input.readUnsignedShort();
         for(var _i2:uint = 0; _i2 < _skinsLen; _i2++)
         {
            _val2 = input.readVarUhShort();
            if(_val2 < 0)
            {
               throw new Error("Forbidden value (" + _val2 + ") on elements of skins.");
            }
            this.skins.push(_val2);
         }
         var _indexedColorsLen:uint = input.readUnsignedShort();
         for(var _i3:uint = 0; _i3 < _indexedColorsLen; _i3++)
         {
            _val3 = input.readInt();
            this.indexedColors.push(_val3);
         }
         var _scalesLen:uint = input.readUnsignedShort();
         for(var _i4:uint = 0; _i4 < _scalesLen; _i4++)
         {
            _val4 = input.readVarShort();
            this.scales.push(_val4);
         }
         var _subentitiesLen:uint = input.readUnsignedShort();
         for(var _i5:uint = 0; _i5 < _subentitiesLen; _i5++)
         {
            _item5 = new SubEntity();
            _item5.deserialize(input);
            this.subentities.push(_item5);
         }
      }

*_bonesIdFunc*

      private function _bonesIdFunc(input:ICustomDataInput) : void
      {
         this.bonesId = input.readVarUhShort();
         if(this.bonesId < 0)
         {
            throw new Error("Forbidden value (" + this.bonesId + ") on element of EntityLook.bonesId.");
         }
      }


readVarUhShort is a customised function from ICustomDataInput which reads the next short and return an uint. Since the uint data type uses 4 bytes to represent an unsigned integer value, then the function should read the next 4 bytes, or next 32 bits (Each byte consists of 8 bits)

In [223]:
def read_var_uh_short(bytes_):
    # read 4 bytes (32 bits)
    
    res_hex = ":".join(bytes_[:4])
    res = convert_byte_to_int(res_hex)
    
    new_bytes_ = bytes_[4:]
    
    return res, new_bytes_

In [224]:
res, bytes_to_read = read_var_uh_short(bytes_to_read)

In [225]:
res

4278452224

**this.bonesId will take the value of res (4278452224 in the example) then the code checks if the value is negative and raises an error if so**

In [226]:
transform_list_to_bytes_array(bytes_to_read)

'00:04:01:dc:14:3c:02:ff:ff:f0:03:ff:ff:f0:04:76:64:43:00:01:78:00:01:02:00:02:00:06:50:d1:10:a8:02:e8:17:fb:1b:f3:03:00:07:01:ff:d3:83:02:ff:ff:ff:03:ff:fe:db:04:ff:ff:ff:05:00:00:00:07:00:00:00:08:ff:ff:ff:00:01:8c:01:00:00:2e:d5:0b:42:66:63:73:00:24:a0:00:00:08:00:3c:21:5c:42:4c:01:b1:80:92:80:00:ff:04:00:01:a8:1a:00:04:01:6a:48:00:02:6a:48:00:03:6a:48:00:04:73:60:42:00:01:78:00:02:06:00:a9:01:00:00:00:00:00:00:00:00:02:00:02:00:06:6e:ff:10:2b:e3:02:e2:02:de:05:00:05:01:ff:da:95:02:2d:2d:2d:03:27:4e:5c:04:23:47:51:05:2f:2f:2f:00:01:91:01:00:00:2e:d5:0b:42:4c:01:b1:80:92:80:00:00:08:00:3c:21:3d:42:66:6c:2b:c0:24:a0:00:8a:13:00:00:00:00:00:00:00:01:02:00:02:00:06:1f:85:10:f4:01:92:01:81:0c:9b:08:00:05:01:ea:c0:62:02:f3:e0:a4:03:00:00:00:04:ff:ff:ff:05:ad:8d:39:00:01:5f:00:00:2e:d5:0b:42:66:6c:2b:c0:24:a0:00:00:08:00'

         var _skinsLen:uint = input.readUnsignedShort();
         for(var _i2:uint = 0; _i2 < _skinsLen; _i2++)
         {
            _val2 = input.readVarUhShort();
            if(_val2 < 0)
            {
               throw new Error("Forbidden value (" + _val2 + ") on elements of skins.");
            }
            this.skins.push(_val2);
         }

In [227]:
def read_unsigned_short(bytes_):
    # 16-bits or 2 bytes
    res_hex = ":".join(bytes_[:2])
    res = convert_byte_to_int(res_hex)

    new_bytes_ = bytes_[2:]

    return res, new_bytes_

In [228]:
skins_len, bytes_to_read = read_unsigned_short(bytes_to_read)

In [229]:
transform_list_to_bytes_array(bytes_to_read)

'01:dc:14:3c:02:ff:ff:f0:03:ff:ff:f0:04:76:64:43:00:01:78:00:01:02:00:02:00:06:50:d1:10:a8:02:e8:17:fb:1b:f3:03:00:07:01:ff:d3:83:02:ff:ff:ff:03:ff:fe:db:04:ff:ff:ff:05:00:00:00:07:00:00:00:08:ff:ff:ff:00:01:8c:01:00:00:2e:d5:0b:42:66:63:73:00:24:a0:00:00:08:00:3c:21:5c:42:4c:01:b1:80:92:80:00:ff:04:00:01:a8:1a:00:04:01:6a:48:00:02:6a:48:00:03:6a:48:00:04:73:60:42:00:01:78:00:02:06:00:a9:01:00:00:00:00:00:00:00:00:02:00:02:00:06:6e:ff:10:2b:e3:02:e2:02:de:05:00:05:01:ff:da:95:02:2d:2d:2d:03:27:4e:5c:04:23:47:51:05:2f:2f:2f:00:01:91:01:00:00:2e:d5:0b:42:4c:01:b1:80:92:80:00:00:08:00:3c:21:3d:42:66:6c:2b:c0:24:a0:00:8a:13:00:00:00:00:00:00:00:01:02:00:02:00:06:1f:85:10:f4:01:92:01:81:0c:9b:08:00:05:01:ea:c0:62:02:f3:e0:a4:03:00:00:00:04:ff:ff:ff:05:ad:8d:39:00:01:5f:00:00:2e:d5:0b:42:66:6c:2b:c0:24:a0:00:00:08:00'

In [166]:
# save = bytes_to_read

In [185]:
# bytes_to_read = save

In [230]:
skins = []
for it in range(0, skins_len):
    uh_short, bytes_to_read = read_var_uh_short(bytes_to_read)
    print(uh_short)
    
    if uh_short < 0:
        print("Error")
        continue
    else:
        skins.append(uh_short) # this.skins.push(_val2)

31200316
50331632
67108848
74867779


In [231]:
transform_list_to_bytes_array(bytes_to_read)

'00:01:78:00:01:02:00:02:00:06:50:d1:10:a8:02:e8:17:fb:1b:f3:03:00:07:01:ff:d3:83:02:ff:ff:ff:03:ff:fe:db:04:ff:ff:ff:05:00:00:00:07:00:00:00:08:ff:ff:ff:00:01:8c:01:00:00:2e:d5:0b:42:66:63:73:00:24:a0:00:00:08:00:3c:21:5c:42:4c:01:b1:80:92:80:00:ff:04:00:01:a8:1a:00:04:01:6a:48:00:02:6a:48:00:03:6a:48:00:04:73:60:42:00:01:78:00:02:06:00:a9:01:00:00:00:00:00:00:00:00:02:00:02:00:06:6e:ff:10:2b:e3:02:e2:02:de:05:00:05:01:ff:da:95:02:2d:2d:2d:03:27:4e:5c:04:23:47:51:05:2f:2f:2f:00:01:91:01:00:00:2e:d5:0b:42:4c:01:b1:80:92:80:00:00:08:00:3c:21:3d:42:66:6c:2b:c0:24:a0:00:8a:13:00:00:00:00:00:00:00:01:02:00:02:00:06:1f:85:10:f4:01:92:01:81:0c:9b:08:00:05:01:ea:c0:62:02:f3:e0:a4:03:00:00:00:04:ff:ff:ff:05:ad:8d:39:00:01:5f:00:00:2e:d5:0b:42:66:6c:2b:c0:24:a0:00:00:08:00'

         var _indexedColorsLen:uint = input.readUnsignedShort();
         for(var _i3:uint = 0; _i3 < _indexedColorsLen; _i3++)
         {
            _val3 = input.readInt();
            this.indexedColors.push(_val3);
         }

In [232]:
indexed_colors_len, bytes_to_read = read_unsigned_short(bytes_to_read)

In [233]:
transform_list_to_bytes_array(bytes_to_read)

'78:00:01:02:00:02:00:06:50:d1:10:a8:02:e8:17:fb:1b:f3:03:00:07:01:ff:d3:83:02:ff:ff:ff:03:ff:fe:db:04:ff:ff:ff:05:00:00:00:07:00:00:00:08:ff:ff:ff:00:01:8c:01:00:00:2e:d5:0b:42:66:63:73:00:24:a0:00:00:08:00:3c:21:5c:42:4c:01:b1:80:92:80:00:ff:04:00:01:a8:1a:00:04:01:6a:48:00:02:6a:48:00:03:6a:48:00:04:73:60:42:00:01:78:00:02:06:00:a9:01:00:00:00:00:00:00:00:00:02:00:02:00:06:6e:ff:10:2b:e3:02:e2:02:de:05:00:05:01:ff:da:95:02:2d:2d:2d:03:27:4e:5c:04:23:47:51:05:2f:2f:2f:00:01:91:01:00:00:2e:d5:0b:42:4c:01:b1:80:92:80:00:00:08:00:3c:21:3d:42:66:6c:2b:c0:24:a0:00:8a:13:00:00:00:00:00:00:00:01:02:00:02:00:06:1f:85:10:f4:01:92:01:81:0c:9b:08:00:05:01:ea:c0:62:02:f3:e0:a4:03:00:00:00:04:ff:ff:ff:05:ad:8d:39:00:01:5f:00:00:2e:d5:0b:42:66:6c:2b:c0:24:a0:00:00:08:00'

In [234]:
def read_int(bytes_):
    # read 4 bytes (32 bits) and returns an int
    
    res_hex = ":".join(bytes_[:4])
    res = convert_byte_to_int(res_hex)
    
    new_bytes_ = bytes_[4:]
    
    return res, new_bytes_

In [235]:
indexed_colors = []

for it in range(0, indexed_colors_len):
    int_, bytes_to_read = read_int(bytes_to_read)
    print(int_)
    
    indexed_colors.append(int_)

2013266178


In [236]:
transform_list_to_bytes_array(bytes_to_read)

'00:02:00:06:50:d1:10:a8:02:e8:17:fb:1b:f3:03:00:07:01:ff:d3:83:02:ff:ff:ff:03:ff:fe:db:04:ff:ff:ff:05:00:00:00:07:00:00:00:08:ff:ff:ff:00:01:8c:01:00:00:2e:d5:0b:42:66:63:73:00:24:a0:00:00:08:00:3c:21:5c:42:4c:01:b1:80:92:80:00:ff:04:00:01:a8:1a:00:04:01:6a:48:00:02:6a:48:00:03:6a:48:00:04:73:60:42:00:01:78:00:02:06:00:a9:01:00:00:00:00:00:00:00:00:02:00:02:00:06:6e:ff:10:2b:e3:02:e2:02:de:05:00:05:01:ff:da:95:02:2d:2d:2d:03:27:4e:5c:04:23:47:51:05:2f:2f:2f:00:01:91:01:00:00:2e:d5:0b:42:4c:01:b1:80:92:80:00:00:08:00:3c:21:3d:42:66:6c:2b:c0:24:a0:00:8a:13:00:00:00:00:00:00:00:01:02:00:02:00:06:1f:85:10:f4:01:92:01:81:0c:9b:08:00:05:01:ea:c0:62:02:f3:e0:a4:03:00:00:00:04:ff:ff:ff:05:ad:8d:39:00:01:5f:00:00:2e:d5:0b:42:66:6c:2b:c0:24:a0:00:00:08:00'

         var _scalesLen:uint = input.readUnsignedShort();
         for(var _i4:uint = 0; _i4 < _scalesLen; _i4++)
         {
            _val4 = input.readVarShort();
            this.scales.push(_val4);
         }

In [237]:
scales_len, bytes_to_read = read_unsigned_short(bytes_to_read)

In [239]:
transform_list_to_bytes_array(bytes_to_read)

'00:06:50:d1:10:a8:02:e8:17:fb:1b:f3:03:00:07:01:ff:d3:83:02:ff:ff:ff:03:ff:fe:db:04:ff:ff:ff:05:00:00:00:07:00:00:00:08:ff:ff:ff:00:01:8c:01:00:00:2e:d5:0b:42:66:63:73:00:24:a0:00:00:08:00:3c:21:5c:42:4c:01:b1:80:92:80:00:ff:04:00:01:a8:1a:00:04:01:6a:48:00:02:6a:48:00:03:6a:48:00:04:73:60:42:00:01:78:00:02:06:00:a9:01:00:00:00:00:00:00:00:00:02:00:02:00:06:6e:ff:10:2b:e3:02:e2:02:de:05:00:05:01:ff:da:95:02:2d:2d:2d:03:27:4e:5c:04:23:47:51:05:2f:2f:2f:00:01:91:01:00:00:2e:d5:0b:42:4c:01:b1:80:92:80:00:00:08:00:3c:21:3d:42:66:6c:2b:c0:24:a0:00:8a:13:00:00:00:00:00:00:00:01:02:00:02:00:06:1f:85:10:f4:01:92:01:81:0c:9b:08:00:05:01:ea:c0:62:02:f3:e0:a4:03:00:00:00:04:ff:ff:ff:05:ad:8d:39:00:01:5f:00:00:2e:d5:0b:42:66:6c:2b:c0:24:a0:00:00:08:00'

In [243]:
def read_var_short(bytes_):
    # read 2 bytes (16 bits) and returns an int
    
    res_hex = ":".join(bytes_[:2])
    res = convert_byte_to_int(res_hex)
    
    new_bytes_ = bytes_[2:]
    
    return res, new_bytes_

In [244]:
scales = []
for it in range(0, scales_len):
    short, bytes_to_read = read_var_short(bytes_to_read)
    print(short)
    
    if short < 0:
        print("Error")
        continue
    else:
        scales.append(short) # this.scales.push(_val2)

6
20689


In [246]:
transform_list_to_bytes_array(bytes_to_read)

'10:a8:02:e8:17:fb:1b:f3:03:00:07:01:ff:d3:83:02:ff:ff:ff:03:ff:fe:db:04:ff:ff:ff:05:00:00:00:07:00:00:00:08:ff:ff:ff:00:01:8c:01:00:00:2e:d5:0b:42:66:63:73:00:24:a0:00:00:08:00:3c:21:5c:42:4c:01:b1:80:92:80:00:ff:04:00:01:a8:1a:00:04:01:6a:48:00:02:6a:48:00:03:6a:48:00:04:73:60:42:00:01:78:00:02:06:00:a9:01:00:00:00:00:00:00:00:00:02:00:02:00:06:6e:ff:10:2b:e3:02:e2:02:de:05:00:05:01:ff:da:95:02:2d:2d:2d:03:27:4e:5c:04:23:47:51:05:2f:2f:2f:00:01:91:01:00:00:2e:d5:0b:42:4c:01:b1:80:92:80:00:00:08:00:3c:21:3d:42:66:6c:2b:c0:24:a0:00:8a:13:00:00:00:00:00:00:00:01:02:00:02:00:06:1f:85:10:f4:01:92:01:81:0c:9b:08:00:05:01:ea:c0:62:02:f3:e0:a4:03:00:00:00:04:ff:ff:ff:05:ad:8d:39:00:01:5f:00:00:2e:d5:0b:42:66:6c:2b:c0:24:a0:00:00:08:00'

In [251]:
test = "ff:ff:ff:05:ad:8d:39:00:01:5f:00:00:2e:d5:0b:42:66:6c:2b:c0:24:a0:00:00:08:00"
len(test.split(":"))

26

         var _subentitiesLen:uint = input.readUnsignedShort();
         for(var _i5:uint = 0; _i5 < _subentitiesLen; _i5++)
         {
            _item5 = new SubEntity();
            _item5.deserialize(input);
            this.subentities.push(_item5);
         }

In [247]:
_subentitiesLen, bytes_to_read = read_unsigned_short(bytes_to_read)

In [248]:
transform_list_to_bytes_array(bytes_to_read)

'02:e8:17:fb:1b:f3:03:00:07:01:ff:d3:83:02:ff:ff:ff:03:ff:fe:db:04:ff:ff:ff:05:00:00:00:07:00:00:00:08:ff:ff:ff:00:01:8c:01:00:00:2e:d5:0b:42:66:63:73:00:24:a0:00:00:08:00:3c:21:5c:42:4c:01:b1:80:92:80:00:ff:04:00:01:a8:1a:00:04:01:6a:48:00:02:6a:48:00:03:6a:48:00:04:73:60:42:00:01:78:00:02:06:00:a9:01:00:00:00:00:00:00:00:00:02:00:02:00:06:6e:ff:10:2b:e3:02:e2:02:de:05:00:05:01:ff:da:95:02:2d:2d:2d:03:27:4e:5c:04:23:47:51:05:2f:2f:2f:00:01:91:01:00:00:2e:d5:0b:42:4c:01:b1:80:92:80:00:00:08:00:3c:21:3d:42:66:6c:2b:c0:24:a0:00:8a:13:00:00:00:00:00:00:00:01:02:00:02:00:06:1f:85:10:f4:01:92:01:81:0c:9b:08:00:05:01:ea:c0:62:02:f3:e0:a4:03:00:00:00:04:ff:ff:ff:05:ad:8d:39:00:01:5f:00:00:2e:d5:0b:42:66:6c:2b:c0:24:a0:00:00:08:00'

In [249]:
_subentitiesLen

4264